### ISSUE TITLE
`warn(needless_borrow)` creates false positive when using `derive(Debug)`

### ISSUE THREAD

**Issue by U007D** [rust-lang-nursery/rust-clippy#2740]
Note: false positive is only emitted when `#[derive(Debug)]` attribute is present:

```
#[derive(Debug)]  // must be present to get false positive
pub struct Crowd<'a> {
    people: &'a [usize],
}
```
gives
```
   Compiling playground v0.0.1 (file:///playground)
warning: this pattern creates a reference to a reference
 --> src/main.rs:7:5
  |
7 |     people: &'a [usize],
  |     ^^^^^^^^^^^^^^^^^^^ help: change this to: `people: &'a [usize]`
  |
note: lint level defined here
 --> src/main.rs:2:9
  |
2 | #![warn(needless_borrow)]
  |         ^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.195/index.html#needless_borrow

warning: this pattern creates a reference to a reference
 --> src/main.rs:7:5
  |
7 |     people: &'a [usize],
  |     ^^^^^^^^^^^^^^^^^^^ help: change this to: `people: &'a [usize]`
  |
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.195/index.html#needless_borrow

    Finished dev [unoptimized + debuginfo] target(s) in 0.66 secs
```
[playground example](http://play.rust-lang.org/?gist=0eb7f3d1b071587cbe063be8a8bb70f7&version=stable&mode=debug)

**Comment by phansch** [comment]
Looks like the macro check in `check_pat` of `needless_borrow` is not working because of https://github.com/rust-lang/rust/pull/49154.

Also: #2651
